Remove use of curl if not needed

krisgesling opened the pull request:
#### Description
This removes the use of Curl to download streams locally within the Skill. Instead the url is passed to the audioservice which is responsible for playing that stream in the appropriate manner.

Fixes #99

This came about when investigating an issue with the FOX News stream. The query params on the url are causing an issue with curl, whilst the audioservice handles playing the url without issue. It is also reducing the complexity of the Skill itself in favor of the audioservice handling playback.

We may need to add https as a supported protocol to the Simple Audioservice and/or set VLC as the default. However I'm yet to hit any issues with this change alone.

#### Type of PR
- [x] Bugfix

#### Testing
This PR should be manually tested as VK does not actually verify that a stream is playing, only that it intends to play it. Would also be good to verify that it is working as expected across the range of systems that people use to run Mycroft.

#### CLA
- [x] Yes

krisgesling committed:
Fix name of GPB allowing settings to work

krisgesling committed:
Only download media url with curl if audioservice cannot handle it

Curl downloads of the media url have been failing and shouldn't be
necessary in most circumstances. This passes the media url directly
to the audioservice for playback unless it uses the https protocol
and there are no audioservices that cannot handle this. Eg the
simple audioservice currently cannot.

Once https support is added to that service, no explicit change
would be required here.

krisgesling committed:
Only download media url with curl if audioservice cannot handle it

Curl downloads of the media url have been failing and shouldn't be
necessary in most circumstances. This passes the media url directly
to the audioservice for playback unless it uses the https protocol
and there are no audioservices that cannot handle this. Eg the
simple audioservice currently cannot.

Once https support is added to that service, no explicit change
would be required here.

forslund commented:
will this work with the simple audioservice backend? it uses mpg123 which doesn't support https...

forslund commented:
Maybe use `self.audioservice.available_backends()` to get the info if there is a backend supporting https and if not issue a `self.log.warning` and fall back to using curl if no backend supports it? Or speak a message that the url can't be played if the protocol isn't supported?

ken-mycroft reviewed (commented):
While I like the idea it seems like this is a breaking change from the perspective that unless a preferred service is provided at run time mp3s delivered via https will no longer work. I have not individually verified the specific things which might break (like different news stations, music skills, etc) but I suspect there will be some. I think this can and should be handled in the audio service but until then I think we might want to reconsider pushing this through.

ken-mycroft commented:
Just checked the existing news skill. It is a common play skill but it spins up its own curl process to play? Anybody know why that is or have any objections to changing this? I think the best idea is to refactor this stuff a bit so the actual curl fall back code lives in the audio service.

JarbasAl commented:
this came up when i tried to add support for catalan news, my suggestions remain the same:
- move this curl stuff into the simple audio backend so it supports https natively without this hacky stuff
- make the default backend vlc


issue for curl #99 
catalan PR #102

ken-mycroft commented:
Agreed. Note Fox news is also failing because the URL returned is not properly decoded by mpg123 but works fine if vlc is used. Of course the audio service using vlc now opens it up to the issues surrounding qml and screen handling in general. A service that plays audio should not need to be aware of whether it has a screen or not.

forslund commented:
:+1: As long as there is a flag to switch of the curl-shenanigans in the simple audioservice if the standard play_mp3 has been replaced by something better.

krisgesling committed:
Add unit test for fetching ABC link

krisgesling committed:
Fix ABC episode fetching.

The url structure is inconsistent hence traversing the page
and finding the download button is a better option.

krisgesling committed:
Change news station from track title to artist title

krisgesling committed:
Change news station from track title to artist title

krisgesling committed:
Improve test steps to check actual state

krisgesling committed:
Improve test steps to check actual state

krisgesling committed:
reduce test duplication

krisgesling committed:
reduce test duplication

krisgesling commented:
Alright I've reworked this a fair bit. This leaves the existing curl behaviour if the url is https and the default audioservice doesn't support that URI. So if we add support for https into the simple audioservice then it will automatically start using that instead.

Confirming that everything was working has been fairly painful with these tests as we weren't actually confirming the state of the system at each Step. Eg "given nothing is playing" would emit a stop command but not guarantee that "nothing was playing". Hence tests could progress beyond a Step that _should_ have failed, or at least wasn't yet succeeding.

Because of the limitations in our current playback services there doesn't seem to be any good event driven way of determining that playback has actually started, only that the service intends to play something. So rather than leave 5 second sleeps throughout the steps, I've added some loops to check until it succeeds or timeout and fail. These are definitely candidates for improvement when we redo the playback services.

krisgesling commented:
I re-read Ake's comment and realised that there's already a mechanism to fallback to another service that supports https, as long as one of them does. 

So updating for that now and it's raised an issue that the backends don't have consistent track_info. By default `artist` will be an empty String but VLC backend only returns an `artists` list. It seems to be only VLC that has this - at least for audio backends in core itself. So I'm accounting for that but we'll revisit the track_info minimum data in the upcoming music sprint.

krisgesling committed:
handle use of VLC during VK tests

krisgesling committed:
handle use of VLC during VK tests

chrisveilleux commented on __init__.py:
```diff
@@ -61,8 +60,8 @@ def tsf():
 
-def gbp():
-    """Custom news fetcher for GBP news."""
+def gpb():
```
change name to something like "fetch_gbp" and put somewhere in comments what the acronym stands for

chrisveilleux commented on __init__.py:
```diff
@@ -61,8 +60,8 @@ def tsf():
-def gbp():
-    """Custom news fetcher for GBP news."""
+def gpb():
+    """Custom news fetcher for GPB news."""
```
Add to docstring why we need a custom news fetcher for GBP news.

chrisveilleux commented on __init__.py:
```diff
@@ -396,6 +374,47 @@ def handle_latest_news(self, message=None, feed=None):
+        )
+        self.now_playing = station_title
+
+    def is_https_supported(self):
```
better as a property?

chrisveilleux commented on __init__.py:
```diff
@@ -396,6 +374,47 @@ def handle_latest_news(self, message=None, feed=None):
+
+    def is_https_supported(self):
+        """Check if any available audioservice backend supports https"""
+        for name, service in self.audioservice.available_backends().items():
```
the `name` variable appears to be unused.  iterate over the dictionary values instead.

chrisveilleux commented on __init__.py:
```diff
@@ -396,6 +374,47 @@ def handle_latest_news(self, message=None, feed=None):
             self.log.info("Traceback: {}".format(traceback.format_exc()))
             self.speak_dialog("could.not.start.the.news.feed")
 
+    def _play_station(self, station_title, station_url, station_image):
```
add args to docstring

chrisveilleux commented on __init__.py:
```diff
@@ -396,6 +374,47 @@ def handle_latest_news(self, message=None, feed=None):
+        return False
+
+    def download_media_file(self, url):
+        """Download a media file and return path to the stream"""
```
Add args and raises to the docstring

chrisveilleux commented on __init__.py:
```diff
@@ -396,6 +374,47 @@ def handle_latest_news(self, message=None, feed=None):
+        self.curl = subprocess.Popen(args)
+        # Check if downloaded file is actually an error page
+        if contains_html(self.STREAM):
+            raise Exception('Could not fetch valid audio file.')
```
We should never raise a base exception. If one of the built-in exception classes don't work, define a new one.  this promotes having less bare except clauses in `try` blocks

chrisveilleux commented on stations/abc.py:
```diff
@@ -12,28 +12,18 @@
-        url = url_temp.format(year, month, hour, day, month)
-
-    return url
+    domain = "https://www.abc.net.au"
```
Why is this custom fetcher in its own file while the other is at the top of the `__init__.py` file?  Also, functions do something so in general they should be named with a verb as the first word.

krisgesling replied:
This was the first, plan was to come back for the rest. Have moved them all and renamed to a consistent `get_station_url()`

chrisveilleux commented on stations/abc.py:
```diff
@@ -12,28 +12,18 @@
+    latest_briefings_url = f"{domain}/radio/newsradio/news-briefings/"
+    soup = BeautifulSoup(urlopen(latest_briefings_url), features='html.parser')
+    result = soup.find(id="collection-grid3")
+    episode_page_link = result.find_all('a')[0]['href']
```
what do `a`, `[0]`, and `href` represent?  can we make this a little less cryptic?

krisgesling replied:
Added comments, which is kind of the best we can do here I think. Do the two comments above make it clear enough?

krisgesling committed:
Create Station classes for all stations.

All stations are now created using one of three classes:
1. FileStation for feeds with a static media file
2. FetchStation for those requiring a custom fetch function
3. RSSStation for those being read from an RSS feed

krisgesling committed:
Create Station classes for all stations.

All stations are now created using one of three classes:
1. FileStation for feeds with a static media file
2. FetchStation for those requiring a custom fetch function
3. RSSStation for those being read from an RSS feed

krisgesling committed:
Extract utils to submodule for cleanliness

krisgesling committed:
Extract utils to submodule for cleanliness

krisgesling committed:
Fix custom station creation

krisgesling committed:
Fix custom station creation

krisgesling committed:
Fix station matching and cleanup

krisgesling committed:
Fix station matching and cleanup

krisgesling committed:
catch explicit request for 'other news'

krisgesling committed:
catch explicit request for 'other news'

krisgesling committed:
remove unused FeedType Enum

krisgesling committed:
remove unused FeedType Enum

krisgesling committed:
Improve docstrings on custom fetcher methods

krisgesling committed:
Improve docstrings on custom fetcher methods

krisgesling committed:
change method to property and simplify

krisgesling committed:
change method to property and simplify

krisgesling committed:
Switch to new standard play handler

krisgesling committed:
Switch to new standard play handler

krisgesling committed:
Fix request by alternate station name

krisgesling committed:
Fix request by alternate station name

krisgesling committed:
Fix and simplify station matching

krisgesling committed:
Fix and simplify station matching

krisgesling committed:
Add args to docstrings

krisgesling committed:
Add args to docstrings

krisgesling committed:
Improve exception specificity

krisgesling committed:
Improve exception specificity

krisgesling committed:
Improve stopping of curl process

- first try to terminate rather than kill
- if process hasn't stopped and provided a return code, then kill
- catch specific exception
- split into dedicated function

krisgesling committed:
Improve stopping of curl process

- first try to terminate rather than kill
- if process hasn't stopped and provided a return code, then kill
- catch specific exception
- split into dedicated function

krisgesling committed:
Reorder Skill functions

1. Startup
2. Intent handlers
3. Remaining methods in alphabetical order

krisgesling committed:
Reorder Skill functions

1. Startup
2. Intent handlers
3. Remaining methods in alphabetical order

krisgesling committed:
Docstring and type hint improvements

krisgesling committed:
Docstring and type hint improvements

krisgesling committed:
Resolve merge conflict

krisgesling committed:
Fix Nonetype has no len() error

chrisveilleux commented on dialog/en-us/PlayTheNews.list:
```diff
@@ -1,6 +1,12 @@
 news
+what's new
+what is new
+whats news
```
Should this be `what's news`?

krisgesling replied:
I'll add some more variations just to be safe.

chrisveilleux commented on stations/__init__.py:
```diff
@@ -11,3 +11,64 @@
+from .tsf import get_tsf_url
+
+# NOTE: This list has to be in sync with the settingsmeta select options.
+stations = dict(
```
All the station related stuff should go in the station package

chrisveilleux replied:
I also think it a good idea to have a data class that, perhaps named `Station` or `NewsStation`,  defines the possible attributes for a station.  it looks like some stations have their own `.py` files.  Maybe all stations should have them with code that builds and returns an instance of the data class?

krisgesling replied:
Just discovered that dataclasses got introduced in Python 3.7 so I'd suggest we hold off using them until 3.6 support has been dropped.

I have moved the list of stations into the station module though.

chrisveilleux commented on stations/abc.py:
```diff
@@ -16,14 +16,19 @@
 
-def abc():
-    """Custom news fetcher for ABC News Australia briefing"""
+def get_abc_url():
```
So this is basically a screen scrape?  If so, perhaps mention that in the comments as they can be fragile, breaking whenever the source updates their webpage.

chrisveilleux commented on stations/gpb.py:
```diff
@@ -0,0 +1,40 @@
+# limitations under the License.
+
+import feedparser
+import re
```
The `re` module is a built-in.  it should be imported before the third party libraries with a blank line between

chrisveilleux commented on stations/match.py:
```diff
@@ -0,0 +1,109 @@
+
+    # Check aliases defined in alt.feed.name.value
+    if aliases:
+        match_confidences += [fuzzy_match(phrase, alias) for alias in aliases]
```
There is a list method called `extend()` that is preferred for this type of operation

chrisveilleux commented on stations/match.py:
```diff
@@ -0,0 +1,109 @@
+    if f"{station.acronym.lower()} {news_keyword}" in phrase:
+        # Eg "play DLF News"
+        match_confidences.append(CONF_HIGH_MATCH)
+    if station.acronym.lower() in phrase:
```
should this be an `elif`?

krisgesling replied:
Yeah good catch.

It wouldn't have produced wrong results as it then takes the max confidence, but it's code that doesn't need to run if the conditional above is True.

Fixed.

chrisveilleux commented on stations/match.py:
```diff
@@ -0,0 +1,109 @@
+        Station or None if news not requested.
+    """
+    match = Match(None, 0.0)
+    if utterance is None:
```
I see logic like this that checks for an absence of an utterance in a lot of places.  What is the use case where an utterance would be empty?  Doesn't activating a skill via the intent parser guarantee an utterance is present?

krisgesling replied:
Not via the Common Play System but this can be refactored to only call this if we know that there is an utterance, rather than checking for an utterance in this function.

It duplicates a little bit of code, but means the `handle_play_request` only has a single way to call it instead of "pass in either an utterance or a station"

chrisveilleux commented on stations/match.py:
```diff
@@ -0,0 +1,109 @@
+    if utterance is None:
+        return match
+
+    # Remove articles like "the" as it matches too well will "other"
```
did you mean `matches too well with`?

chrisveilleux commented on stations/match.py:
```diff
@@ -0,0 +1,109 @@
+    # Remove articles like "the" as it matches too well will "other"
+    search_phrase = utterance.lower().replace('the', '').strip()
+
+    if not skill.voc_match(search_phrase, 'News'):
```
Again, can we really reach this point if the user is not asking for the news?

krisgesling replied:
Yeah via the Common Play Service - so makes sense to move this check to that specific code path.

chrisveilleux commented on stations/station.py:
```diff
@@ -0,0 +1,122 @@
+        self.full_name = full_name
+        self.image_file = image_file
+
+    def as_dict(self):
```
If a skill attribute data class was defined, this could go in there.

chrisveilleux commented on stations/station.py:
```diff
@@ -0,0 +1,122 @@
+        skill_path = Path(__file__).parent.parent.absolute()
+        file_path = Path(skill_path, 'images', self.image_file)
+        if not file_path.exists():
+            LOG.error(
```
Is this really an error condition?  There is a default image.  This feels like a warning or info level message.

krisgesling replied:
good call

chrisveilleux commented on stations/station.py:
```diff
@@ -0,0 +1,122 @@
+        return media_url
+
+    @staticmethod
+    def get_media_from_rss(rss_url):
```
No docstring here.  Why is this a `staticmethod` that passes a `rss_url` instead of using the `self._rss_url` attribute or the `self.media_uri` property?

krisgesling replied:
It gets used by `set_custom_station()` to test if the given url is an rss feed. But I can already hear you saying that we should create a separate function for this - it didn't seem so bad until I typed the first sentence...

chrisveilleux commented on stations/tsf.py:
```diff
@@ -0,0 +1,39 @@
+# limitations under the License.
+
+import requests
+from datetime import timedelta
```
The `datetime` package is a builtin.  It should appear before the third party imports with a blank line between.

chrisveilleux commented on stations/tsf.py:
```diff
@@ -0,0 +1,39 @@
+    feed = ('https://www.tsf.pt/stream/audio/{year}/{month:02d}/'
+            'noticias/{day:02d}/not{hour:02d}.mp3')
+    uri = None
+    i = 0
```
rename to something more meaningful, like `hours_offset`.

chrisveilleux commented on stations/tsf.py:
```diff
@@ -0,0 +1,39 @@
+            'noticias/{day:02d}/not{hour:02d}.mp3')
+    uri = None
+    i = 0
+    status = 404
```
Use the built-in `HTTPStatus` to give meaning to the status codes.

chrisveilleux reviewed (changes requested):
Mostly a good refactor.  I have some suggestions for arranging things a bit differently and defining a data class for news feed attributes.

krisgesling committed:
update docstring to explicitly state it's a scraper

krisgesling committed:
Clean up imports

krisgesling committed:
Add additional variations

krisgesling committed:
Use list.extend rather than += concatenation

krisgesling committed:
Lower confidences not needed if conditional above is true

krisgesling committed:
Simplify handle_play_request

It now has a single way to call it - passing in the station to be played
rather than an either/or with an utterance.

krisgesling committed:
Refactor restarting previously played station

krisgesling committed:
Fix comment typo

krisgesling committed:
Only check for news vocab when called via CPS

Presume that any direct intents are requesting the news.

krisgesling committed:
log warning if no station image can be found

krisgesling committed:
Minor clean up of TSF

- Reorder imports
- Use more meaningful variable names
- Use HTTPStatus codes rather than ints

krisgesling committed:
Refactor method to perform single purpose

krisgesling committed:
clean up imports

krisgesling committed:
Move list of stations to station module

krisgesling committed:
Fix failing Play Fox News test

krisgesling committed:
Move country-based default stations from translatable resource

These shouldn't need to be translated so only adds complexity and
confusion. They are now defined in the station module.
